### ISSUE TITLE
jeromq is adding 7 bytes when frame size is large(~MBs) on Android

### ISSUE THREAD

**Issue by yagnasrinath** [zeromq/jeromq#638]
@trevorbernard @markif 
When the message has multiple frames, and if any of the frame size is larger(I noticed it when I am sending around 4MB). It is adding exactly 7 bytes. Although Msg class keeps track of size, but doesn't limit the buffer size before passing it on to ZFrame. I am using a workaround by calling recv and building ZMsg myself.
<del>
```
 private zmq.Msg recv(ZMQ.Socket socket, int flags)
    {
        zmq.Msg msg = socket.base().recv(flags);

        if (msg != null) {
            return msg;
        }
        return null;
    }

    private ZMsg recvMsg(ZMQ.Socket socket, int flag)
    {
        if (socket == null) {
            throw new IllegalArgumentException("socket is null");
        }

        ZMsg msg = new ZMsg();

        while (true) {
            zmq.Msg internalMsgFrame = recv(socket, flag);
            ByteBuffer buf = internalMsgFrame.buf();
            if (buf == null) {
                return null;
            }
            if (internalMsgFrame == null) {
                // If receive failed or was interrupted
                msg.destroy();
                msg = null;
                break;
            }
            // check size and resize;
            buf.limit(internalMsgFrame.size());
            byte[] data = new byte[buf.remaining()];
            buf.duplicate().get(data);
            msg.addLast(data);
            if (!(socket.base().getSocketOpt(zmq.ZMQ.ZMQ_RCVMORE) == 1)) {
                break;
            }
        }
        return msg;
    }
```
</del>
**NOTE:** The above workaround is wrong as android adds 4bytes in beginning and 3 bytes at end.

**Comment by yagnasrinath** [comment]
This seems to be happening because of use of DirectByteBuffer when message size is more than 1MB on Android. DirectByteBuffer seems to allocate extra 7bytes on android(https://android.googlesource.com/platform/libcore/+/android-6.0.1_r21/luni/src/main/java/java/nio/ByteBuffer.java#68) than requested capacity. So when .array() method is called on the bytebuffer, it returns an array with extra seven bytes.

**Comment by fredoboulo** [comment]
That's ... unfortunate. Thanks for pointing out the problem. I don't have Android, so I cannot test the quickfix I made. Would you please tell me if it's headed to the right direction?

**Comment by yagnasrinath** [comment]
Thanks, let me test it.

**Comment by yagnasrinath** [comment]
@fredoboulo Sorry, I tried your fix, but I am having a lot of crashes, when I am trying to create sockets using snapshot from your branch. Unfortunately, I am occupied a bit. I will see if I can fix those issues and report back. I don't see any issue with your PR as such though.

**Comment by yagnasrinath** [comment]
@fredoboulo Also, I found out that Android adds 4 bytes in the beginning and 3 bytes at the end. So I think it is better to identify the vm and then based on it, get the buffer.

**Comment by fredoboulo** [comment]
That's even weirder. I'm afraid that without more infos and direct access to Android environment, I cannot do much on my own.
I'm still interested by any feedback about my branch, the only difference with master is the fixing commit. Interested because if you experience crashes, we might have to work on jeromq master itself.
If you find a fix, you are invited to submit a PR to the project, it could help many others.

**Comment by yagnasrinath** [comment]
May be the easiest fix for android is to default to heap allocation. I will work on it when I can squeeze in some time. Thanks for all the help though.

**Comment by yagnasrinath** [comment]
This seems to happen even when I align to 8bytes. I tried reducing the heap allocation limit to 96, and tried allocating 104 byte frame, It allocated a bytebuffer of size 111. So, my guess is that it allocates the additional 7bytes all the time.

**Comment by fredoboulo** [comment]
Have you tried with other versions of Android? So it seems that it allocates more than the actual capacity to provide alignment, but the array() should not return the extra bytes allocated for that purpose. I'm curious if other releases experience the same behaviour.

**Comment by yagnasrinath** [comment]
According to the source code, we should never access DirectByteBuffer backing array directly using getArray() https://android.googlesource.com/platform/libcore/+blame/master/ojluni/src/main/java/java/nio/ByteBuffer.java @fredoboulo what do you think?

**Comment by fredoboulo** [comment]
I did not see that reference to avoid using getArray(), but there is this [arrayOffset()](https://docs.oracle.com/javase/7/docs/api/java/nio/ByteBuffer.html#arrayOffset()) method that the code could take advantage of, maybe in Msg class when [creating the data array](https://github.com/zeromq/jeromq/blob/master/src/main/java/zmq/Msg.java#L133) or when giving [public access](https://github.com/zeromq/jeromq/blob/master/src/main/java/zmq/Msg.java#L214). I just don't have time to investigate now, but I will get back to the topic when I get the chance (worst case in 2+ weeks).

**Comment by yagnasrinath** [comment]
What I meant was that array()(used in Msg.java) method tries to return **_hb_**, which should not have been set in case of DirectByteBuffer. I feel that Android's implementation is messed up. https://android.googlesource.com/platform/libcore/+blame/master/ojluni/src/main/java/java/nio/ByteBuffer.java#773

Since MappedByteBuffer is super class of DirectByteBuffer, which upon calling super class constructor tries to sets the **_hb_** variable to **_buf_**. This is the cause of all problems. If we do not call array(), but use the get* methods to fetch the data, everything seems to work fine, and I feel it is safer not to access backing array ever.

**Comment by knuesel** [comment]
Is there an update on this? Android is behaving correctly I think: the API doesn't guarantee anything about the arrayOffset for buffers made with `ByteBuffer.allocateDirect`.

I can try to implement a fix if you're not working on it anymore... 

@yagnasrinath : The problem isn't with `hb`: DirectByteBuffer calls the MappedByteBuffer constructor with a buffer (memoryRef.buffer) as 5th argument:
https://android.googlesource.com/platform/libcore/+blame/master/ojluni/src/main/java/java/nio/DirectByteBuffer.java#98
This is passed to the ByteBuffer constructor, which sets the `hb` field. But it is true that there would be no problem if the array was only accessed through the `get` method. But I suppose it would be bad for performance to make unnecessary copies of the array.

**Comment by yagnasrinath** [comment]
True. But the array allocated has extra bytes for padding. It has 7 extra bytes to be precise. The hasArray() should have returned false, disallowing direct access to array using array(). Please feel free to take it forward. I can review your fix if you would like to.

**Comment by knuesel** [comment]
Before doing any change, I checked the JeroMQ codebase for similar problems.

**TL;DR:** no problems founds outside of `zmq.Msg`. However in some places the correctness depends on the caller passing a "good" buffer, so I propose a few changes to make the code more robust.

For `zmq.Msg`: there are problems beyond the Android bug: the internal `byte[]` in `Msg.data` leads to inconsistent behavior. For example, the code

```
import zmq.Msg;

public class Test {
    public static void main(String[] args) {

        Msg msg1 = new Msg(new byte[] {71});

        // Make copy of msg1
        Msg msg2 = new Msg(msg1);

        // Overwrite msg2 first byte
        // This actually writes to the backing array shared by msg1.buf and msg2.buf
        msg2.put(5);

        // Retrieve the byte: looks fine
        System.out.printf("msg1: %d, msg2: %d\n", msg1.get(), msg2.get());

        // Retrieve the same byte with getBytes(int index, byte[] dst, int off, int len)
        // Gives the old value: msg2.data is a stale copy made at construction time.
        byte[] data = new byte[1];
        msg2.getBytes(0, data, 0, 1);
        System.out.printf("msg2 with getBytes(int, byte[], int, int): %d\n", data[0]);
    }
}
```
will print
```
msg1: 5, msg2: 5
msg2 with getBytes(int, byte[], int, int): 71
```
which is obviously wrong.

I propose to fix both the Android bug and the inconsistent behavior by 1) getting rid of the `Msg.data` member, 2) updating the documentation and 3) fix `getBytes(int, byte[], int, int)` to use `ByteBuffer.array()` correctly.

### The details

Excluding the tests, there are 14 uses of `ByteBuffer.array()`. In 6 cases there is no problem: the buffer is created with `ByteBuffer.allocate()` in the same method so the `array()` call is obviously correct.  The remaining cases are:

---

`org.zeromq.ZBeacon.BroadcastServer.handleMessage(ByteBuffer, int, InetAddress)`:

    listener.get().onBeacon(from, Arrays.copyOf(buffer.array(), size));

Not a problem *currently*: it's a private method called only from `ZBeacon.BroadcastServer.run()` with buffer allocated through `ByteBuffer.allocate`. But it's not robust: someone could change the `run` code to use a direct buffer.

**Suggested fix**: replace `array()` call with explicit `byte[]` allocation and `get()`.

---

Six methods of `zmq.io.mechanism.curve.Curve`. These are non-public methods called in `zmq.io.mechanism.curve`.

For example:

    int afternm(ByteBuffer ciphered, ByteBuffer plaintext, int length, ByteBuffer nonce, byte[] precom)
    {
        return afternm(ciphered.array(), plaintext.array(), length, nonce.array(), precom);
    }

    int afternm(byte[] ciphered, byte[] plaintext, int length, byte[] nonce, byte[] precomp)
    {
        return curve25519xsalsa20poly1305.crypto_box_afternm(ciphered, plaintext, length, nonce, precomp);
    }

The buffers passed to these functions are all from `ByteBuffer.allocate`, so it is currently correct, but not robust.

Note that each method comes in two overloads: the overload with `ByteBuffer` arguments is the one used by the Curve code; it simply delegates to the overload with `byte[]` arguments.

**Suggested fix**: remove the overloads with `ByteBuffer` arguments (and change the caller code to use `.array()` there, where it is known to be safe).

---

`zmq.Msg(final ByteBuffer)`: this is our bug. See below:

### The `Msg` case

The class exposes a `byte[]` through `data()` and a `ByteBuffer` through `buf()`. Even without considering our bug, the code is wrong because the `data` and `buf` members can represent different byte arrays:

Case 1: Make a message with

    public Msg(final ByteBuffer src)
    {
        if (src == null) {
            throw new IllegalArgumentException("ByteBuffer cannot be null");
        }
        this.type = Type.DATA;
        this.flags = 0;
        this.buf = src.duplicate();
        if (buf.hasArray() && buf.position() == 0 && buf.limit() == buf.capacity()) {
            this.data = buf.array();
        }
        else {
            this.data = null;
        }
        this.size = buf.remaining();
    }

and suppose we're in the `this.data = null` case. Now we call `data()`:

    public byte[] data()
    {
        if (data == null) {
            data = new byte[buf.remaining()];
            buf.duplicate().get(data);
        }
        return data;
    }

So `data` will contain a *copy* of the `buf` data.

Case 2: Make a message with

    public Msg(final Msg m)
    {
        if (m == null) {
            throw new IllegalArgumentException("Msg cannot be null");
        }
        this.type = m.type;
        this.flags = m.flags;
        this.size = m.size;
        this.buf = m.buf != null ? m.buf.duplicate() : null;
        if (m.data != null) {
            this.data = new byte[this.size];
            System.arraycopy(m.data, 0, this.data, 0, m.size);
        }
    }

If `m` has a backing array, `this.data` will get a copy but `this.buf` will use the original.

In both cases, the behavior of `Msg` will be inconsistent: most methods (including `getBytes(int, ByteBuffer, int)`) will use the data behind `buf`, but `getBytes(int, byte[], int, int)` will use `data`.

Note that only `data()` and `getBytes(int, byte[], int, int)` use the `data` field.

**Suggested fix**: 
1. Remove the private `data` member.
2. Specify in the `Msg.data()` documentation that the returned array might be a copy, so using this method can incur a performance penalty and it cannot reliably be used to change the message data.
3. Change `getBytes(int, byte[], int, int)` to use `arrayOffset` when necessary (we could also use a `get` but I think the performance would be worse for buffers from `ByteBuffer.allocate`... At least OpenJDK uses a loop of `get()` on single bytes).

---

### Appendix

For completeness, here's everywhere `ByteBuffer.allocateDirect` is used in JeroMQ:

`zmq.msg.MsgAllocatorDirect.allocate`: that's what triggers our bug.

`zmq.io.coder.DecoderBase`, member `buf`. Used safely, but exposed through the `getBuffer()` method.

`zmq.io.coder.EncoderBase`, member `buffer`. Used safely, but returned to the caller through the first argument of `encode(ValueReference<ByteBuffer>, int)`.

`zmq.io.coder.raw.RawDecoder`: similar to `zmq.io.coder.DecoderBase`.

I think all these cases are OK: if the caller of `getBuffer()` or `encode(ValueReference<ByteBuffer>, int)` uses `array()`, it's their responsibility to do it correctly.

---

@yagnasrinath why should hasArray() return false? A backing array is available so returning true is correct... The problem is that JeroMQ assumes the array offset is zero, which is not always true: With `ByteBuffer.allocate` the offset will be zero, but with `ByteBuffer.allocateDirect` it is not specified, so the user must check `arrayOffset`:
https://docs.oracle.com/javase/7/docs/api/java/nio/ByteBuffer.html#allocateDirect(int)
